dokka-integration-tests/gradle/src/main/kotlin/org/jetbrains/dokka/it/gradle/junit/GradleTestProjectInitializer.kt

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2014-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+package org.jetbrains.dokka.it.gradle.junit
+
+import org.jetbrains.dokka.it.gradle.AbstractGradleIntegrationTest.Companion.updateProjectLocalMavenDir
+import org.jetbrains.dokka.it.gradle.junit.DokkaGradlePluginTestExtension.Companion.getAndroidSdkDir
+import java.nio.file.Path
+import kotlin.io.path.*
+
+/**
+ * Initialise a Gradle Test project into a directory.
+ *
+ * The source directory must not be modified.
+ */
+fun interface GradleTestProjectInitializer {
+
+    /**
+     * Must copy the contents of [source] into [destination].
+     *
+     * Should create a `gradle.properties` file with the contents of [gradleProperties].
+     */
+    fun initialize(
+        source: Path,
+        destination: Path,
+        testedVersions: TestedVersions,
+        gradleProperties: Map<String, String>,
+    )
+
+    object Default : GradleTestProjectInitializer {
+        override fun initialize(
+            source: Path,
+            destination: Path,
+            testedVersions: TestedVersions,
+            gradleProperties: Map<String, String>,
+        ) {
+            require(source.isDirectory()) { "Source project must be a directory. Actual: $source" }
+            source.copyToRecursively(destination, overwrite = true, followLinks = false)
+
+            destination.updateProjectLocalMavenDir()
+
+            createGradlePropertiesFile(destination, gradleProperties)
+
+            updateVersions(destination, testedVersions)
+
+            if (testedVersions.agp != null) {
+                initialiseAndroid(
+                    destination = destination
+                )
+            }
+        }
+
+        private fun createGradlePropertiesFile(
+            projectDir: Path,
+            gradleProperties: Map<String, String>,
+        ) {
+            val gradlePropertiesFile = projectDir.resolve("gradle.properties").apply {
+                deleteIfExists()
+                createFile()
+            }
+
+            gradlePropertiesFile.writeText(
+                gradleProperties.entries
+                    .map { (k, v) -> "$k=$v" }
+                    .sorted()
+                    .joinToString(separator = "\n", postfix = "\n")
+            )
+        }
+
+        private fun updateVersions(
+            projectDir: Path,
+            testedVersions: TestedVersions,
+        ) {
+            projectDir.walk().filter { it.isRegularFile() }.forEach { file ->
```

Comment:
Curious, if this will not become a performance issue.
AFAIU we are here iterating through ALL files in the project, reading and writing it.
F.e in `ui-showcase` we have versioning with 5 versions, and so 5 Dokka’s HTML outputs :)

Of course for now it should not be an issue, but still, I think that we could at limit those modifications to a list of well known files (build.gradle.kts, settings.gradle.kts, libs.versions.toml)

Reply:
Good idea, I'll add a filter.